Google Play services crashing when missing "Nearby devices" permission and using Bluetooth based passkey authentication

When using Bluetooth based passkey authentication e.g. with Microsoft Authenticator, the "Nearby devices" permission is required by Google Play Services.

I'm not sure if this can be caught by the compatibility layer, but a prompt to ask for the permission would improve the user experience.

Device: Pixel 9 Pro
GrapheneOS build: 2025073000
Google Play services: 25.26.35 (260400-73060121)
GmsCompatConfig: 160

```
07-28 09:50:16.770  7992  7992 E AndroidRuntime: FATAL EXCEPTION: main
07-28 09:50:16.770  7992  7992 E AndroidRuntime: Process: com.google.android.gms.ui, PID: 7992
07-28 09:50:16.770  7992  7992 E AndroidRuntime: java.lang.SecurityException: Permission Denial: starting Intent { act=android.bluetooth.adapter.action.REQUEST_ENABLE xflg=0x4 cmp=com.android.settings/.bluetooth.RequestPermissionActivity } from ProcessRecord{f01400a 7992:com.google.android.gms.ui/u10a189} (pid=7992, uid=1010189) requires android.permission.BLUETOOTH_CONNECT
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:3349)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:3333)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:3309)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:3251)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.app.IActivityTaskManager$Stub$Proxy.startActivity(IActivityTaskManager.java:2178)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.app.Instrumentation.execStartActivity(Instrumentation.java:2006)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.app.Activity.startActivityForResult(Activity.java:6043)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wft.platform_startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at augf.platform_startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):4)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wfs.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.google.android.chimera.android.Activity.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wjv.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wfq.support_startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wfv.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wkt.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):4)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wfv.public_startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):2)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wft.startActivityForResult(:com.google.android.gms@252635035@25.26.35 (260400-783060121):3)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at wkq.a(:com.google.android.gms@252635035@25.26.35 (260400-783060121):212)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at acv.b(:com.google.android.gms@252635035@25.26.35 (260400-783060121):26)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at dj.b(:com.google.android.gms@252635035@25.26.35 (260400-783060121):11)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at bkvs.run(:com.google.android.gms@252635035@25.26.35 (260400-783060121):19)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at bkww.b(:com.google.android.gms@252635035@25.26.35 (260400-783060121):8)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at bkvq.onClick(:com.google.android.gms@252635035@25.26.35 (260400-783060121):18)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.view.View.performClick(View.java:8083)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.google.android.material.button.MaterialButton.performClick(:com.google.android.gms@252635035@25.26.35 (260400-783060121):16)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.view.View.performClickInternal(View.java:8060)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.view.View$PerformClick.run(View.java:31549)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:995)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:103)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:248)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:338)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:9106)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:593)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.internal.os.ExecInit.main(ExecInit.java:50)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:371)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: Caused by: android.os.RemoteException: Remote stack trace:
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.server.wm.ActivityTaskSupervisor.checkStartAnyActivityPermission(ActivityTaskSupervisor.java:1211)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.server.wm.ActivityStarter.executeRequest(ActivityStarter.java:1239)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:869)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1315)
07-28 09:50:16.770  7992  7992 E AndroidRuntime: 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1259)
```

--- a/core/api/system-current.txt
+++ b/core/api/system-current.txt
@@ -2272,10 +2272,6 @@ package android.app.compat.gms {
     method public static boolean isEnabledFor(@NonNull String, int);
   }
 
-  public interface GmsCorePackageFlag {
-    field public static final int GRANT_PERMS_FOR_ICC_AUTHENTICATION = 0; // 0x0
-  }
-
   public class GmsUtils {
     method @NonNull public static android.content.Intent createAppPlayStoreIntent(@NonNull String);
   }

--- a/core/java/android/app/compat/gms/GmsCorePackageFlag.java
+++ b/core/java/android/app/compat/gms/GmsCorePackageFlag.java
@@ -1,9 +1,6 @@
 package android.app.compat.gms;
 
-import android.annotation.SystemApi;
-
 /** @hide */
-@SystemApi
 public interface GmsCorePackageFlag {
     int GRANT_PERMS_FOR_ICC_AUTHENTICATION = 0;
 }

--- a/core/java/com/android/internal/gmscompat/util/GmcActivityUtils.java
+++ b/core/java/com/android/internal/gmscompat/util/GmcActivityUtils.java
@@ -121,6 +121,17 @@ public class GmcActivityUtils implements Application.ActivityLifecycleCallbacks 
             return null;
         }
 
+        if (BluetoothAdapter.ACTION_REQUEST_ENABLE.equals(intent.getAction())) {
+            if (GmsCompat.isGmsCore() && !GmsCompat.hasPermission(Manifest.permission.BLUETOOTH_CONNECT)) {
+                try {
+                    GmsCompatApp.iGms2Gca().showGmsCoreMissingNearbyDevicesPermissionGeneric();
+                } catch (RemoteException e) {
+                    GmsCompatApp.callFailed(e);
+                }
+                return null;
+            }
+        }
+
         String pkg = cn != null ? cn.getPackageName() : intent.getPackage();
         if (pkg != null) {
             boolean checkIntent = false;

--- a/services/core/java/com/android/server/pm/PackageInstallerSession.java
+++ b/services/core/java/com/android/server/pm/PackageInstallerSession.java
@@ -4979,7 +4979,7 @@ public class PackageInstallerSession extends IPackageInstallerSession.Stub {
                     case PackageId.PIXEL_HEALTH_NAME:
                         throw new PackageManagerException(errorCode,
                                 "Only the first-party package source and shell are allowed " +
-                                        "to install " + PackageId.PIXEL_HEALTH_NAME);
+                                        "to install " + mPackageName);
                 }
             }
         }

